pkg/api/validation/validation.go

```diff
@@ -555,15 +555,50 @@ func validateEnv(vars []api.EnvVar) errs.ValidationErrorList {
 		vErrs := errs.ValidationErrorList{}
 		if len(ev.Name) == 0 {
 			vErrs = append(vErrs, errs.NewFieldRequired("name"))
-		}
-		if !util.IsCIdentifier(ev.Name) {
+		} else if !util.IsCIdentifier(ev.Name) {
 			vErrs = append(vErrs, errs.NewFieldInvalid("name", ev.Name, cIdentifierErrorMsg))
 		}
+		vErrs = append(vErrs, validateEnvVarSources(ev).Prefix("source")...)
 		allErrs = append(allErrs, vErrs.PrefixIndex(i)...)
 	}
 	return allErrs
 }
 
+func validateEnvVarSources(ev api.EnvVar) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if ev.Source == nil {
+		return allErrs
+	}
+
+	numSources := 0
+
+	switch {
+	case ev.Source.PodField != nil:
+		numSources++
+		allErrs = append(allErrs, validateObjectFieldSelector(ev.Source.PodField).Prefix("podField")...)
+	}
+
+	if ev.Value != "" && numSources != 0 {
+		allErrs = append(allErrs, errs.NewFieldInvalid("", "", "sources cannot be specified when value is not empty"))
+	}
+
+	return allErrs
+}
+
+func validateObjectFieldSelector(fs *api.ObjectFieldSelector) errs.ValidationErrorList {
```

Comment:
@thockin At this point, that happens in the kubelet.  I don't think that's great.  Let me think on this a bit more.